doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,617 @@
+
+# 2 Functionality
+
+## 2.1 Target Deployment Use Cases
+
+The `apply-patch` method should help with automating partial config updates, as external systems can generate the update patch, and apply.
+
+The `checkpoint` and `rollback` commands should help improve recoverability, can also be used by external systems to help revert failures during `apply-patch` operation.
+
+Human operators can also leverage the `checkpoint` and `rollback` functionalities while doing updates through the CLI using [SONiC CLI commands](https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md).
+
+## 2.2 Functional Description
+
+### 2.2.1 Apply-Patch
+The SONiC `apply-patch` command can broadly classified into the following steps
+
+#### Stage-1 Patch Ordering
+As per the JSON standard, the elements described in JSON format are unordered. However, in SONiC there are some tables which can not be processed before another table is updated. For example changes to the VLAN_MEMBER table needs to be processed before the PORTCHANNEL table is processed. If the ordering is not established, the Port Channel delete operation will fail with an error message complaining that the port-channel is still part of a VLAN. So it is essential that the dependency between the tables is identified and the patch is processed in that order.
+
+The `apply-patch` command uses an ordering table which specifies the list of dependent tables that need to be processed before a table is processed. Below is an example depiction of two ConfigDB tables.
+
+```
+{
+  "VLAN" : ["VLAN_MEMBER", "VLAN_INTERFACE"],
+  "PORTCHANNEL" : ["VLAN_MEMBERSHIP", "PORTCHANNEL_MEMBER"]
+}
+```
+A dependency graph is created for all the tables which are in the JSON patch file. The dependency graph is then resolved to find the order in which the updates to tables in the JSON patch are performed. The JSON patch file is then sorted in the resolved order.
+
+#### Stage-2 Identifying Services to Restart
+There are a few SONiC applications which store its configuration in the ConfigDB. These applications do not subscribe to the ConfigDB change events. So any changes to their corresponding table entries as part of the patch apply process in the ConfigDB are not processed by the application immediately. In order to apply the configuration changes, corresponding service needs to be restarted. A list of such ConfigDB tables and corresponding systemd service is specified in the *apply-patch* command.  See below for an example.
+
+```
+{
+  "SYSLOG_SERVER": ["rsyslog"],
+  "DHCP_SERVER": ["dhcp_relay"],
+  "NTP_SERVER": ["ntp-config.service", "ntp.service"]
+  "BGP_MONITORS": ["bgp"],
+  "BUFFER_PROFILE": ["swss"],
+  "RESTAPI": ["restapi"]
+}
+```
+
+This table is used to restart corresponding systemd service when a patch operation has been perform on its corresponding ConfigDB table. 
+
+#### Stage-3 Add delay after table updates for listening services to absorb changes
```

Comment:
>Add delay after table updates for listening services to absorb changes [](start = 13, length = 70)

Offline discussed, delay is not a stable way to check changes absorbed. Let's depend on 'service restart' response, or use cli to verify it.